logstash_async/transport.py

```diff
@@ -3,15 +3,155 @@
 # This software may be modified and distributed under the terms
 # of the MIT license.  See the LICENSE file for details.
 
+from abc import ABC, abstractmethod
+from time import sleep
+from typing import Union
+import json
+import random
 import socket
 import ssl
 import sys
 
+from requests.auth import HTTPBasicAuth
 import pylogbeat
+import requests
 
 from logstash_async.utils import ichunked
 
 
+class Transport(ABC):
+    """The :class:`Transport <Transport>` is the abstract base class of
+    all transport protocols.
+
+    :param host: The name of the host
+    :type host: str
+    :param port: The port number of the service
+    :type port: int
+    :param timeout: The timeout for the connection
+    :type timeout: float
+    :param codec: The codec which will be used to de/encode the data (eg. plain
+    or json)
+    :type codec: str
+    :param ssl_enable: Use TLS for the transport (Default: True)
+    :type ssl_enable: bool
+    :param ssl_verify: If True the class tries to verify the TLS certificate
+    with certifi. If you pass a string with a file location to CA certificate
+    the class tries to validate it against it. (Default: True)
+    :type ssl_verify: bool or str
+    """
+
+    def __init__(
+            self,
+            host: str,
+            port: int,
+            timeout: float,
+            codec: str,
+            ssl_enable: bool = True,
+            ssl_verify: bool = True
+    ) -> None:
+        self.host = host
+        self.port = port
+        self.timeout = timeout
+        self.codec = codec
+        self.ssl_enable = ssl_enable
+        self.ssl_verify = ssl_verify
+        super().__init__()
+
+    @property
+    def host(self) -> str:
+        """host
+        :param name: The name of the host
+        :type name: str
+        :return: The current name of the host
+        :rtype: str
+        """
+        return self.__host
```

Comment:
Is there a reason for using the double score name mangling everywhere?
So far I can't see it's necessary.

If not, I would prefer normal single underscores to mark instance attributes as "semiprivate".

Reply:
I know that there is no real visibility for attributes in python. Anyway I try to keep the attributes "private" not just "protected" if I use properties. From the user perspective, it's more clear for me.

_"Protected"_
```
>>> from logstash_async.transport import HttpTransport
>>> test = HttpTransport('127.0.0.1', '5959')
>>> test.host
'127.0.0.1'
>>> test._host
'127.0.0.1'
>>> test._host = '127.0.0.2'
>>> test.host
'127.0.0.2'
```

_"Private"_
```
>>> from logstash_async.transport import HttpTransport
>>> test = HttpTransport('127.0.0.1', '5959')
>>> test.host
'127.0.0.1'
>>> test.__host
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'HttpTransport' object has no attribute '__host'
>>> test._Transport__host
'127.0.0.1'
>>> test._Transport__host = '127.0.0.2'
>>> test.host
'127.0.0.2'
```

Reply:
Ok, fine by me.